src/systems/label/Label.cc

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#include "Label.hh"
+#include <string>
+
+#include <ignition/plugin/Register.hh>
+#include "ignition/gazebo/EntityComponentManager.hh"
+#include "ignition/gazebo/components/SemanticLabel.hh"
+#include "ignition/gazebo/components/Link.hh"
+#include "ignition/gazebo/components/Visual.hh"
+
+using namespace ignition;
+using namespace gazebo;
+using namespace systems;
+
+//////////////////////////////////////////////////
+Label::Label() : System()
+{
+}
+
+//////////////////////////////////////////////////
+Label::~Label() = default;
+
+//////////////////////////////////////////////////
+void Label::Configure(const Entity &_entity,
+    const std::shared_ptr<const sdf::Element> &_sdf,
+    gazebo::EntityComponentManager &_ecm,
+    gazebo::EventManager & /*_eventMgr*/)
+{
+  const std::string labelTag = "label";
+
+  if (!_sdf->HasElement(labelTag))
+  {
+    ignerr << "Failed to load Label system; label tag not found.\n";
+    return;
+  }
+
+  auto label = _sdf->Get<int>(labelTag);
+
+  if (label < 0 || label > 255)
+  {
+    ignerr << "Failed to configure Label system; value " << label
+      << " is not in [0-255] range.\n";
+    return;
+  }
+
+  const std::string parentName = _sdf->GetParent()->GetName();
+
+  // Set the component to the visual to set its user data, but
+  // if the plugin is inside the <model> tag, get its visual child
+  if (parentName == "visual")
+  {
+    _ecm.CreateComponent(_entity, components::SemanticLabel(label));
+  }
+  else if (parentName == "model")
```

Comment:
Instead of checking the parent SDF for hardcoded strings like `visual` or `model`, you should be able to check if `_entity` is a `visual` or `model` by seeing if it has a [`components::Visual`](https://github.com/ignitionrobotics/ign-gazebo/blob/main/include/ignition/gazebo/components/Visual.hh#L32) or [`components::Model`](https://github.com/ignitionrobotics/ign-gazebo/blob/main/include/ignition/gazebo/components/Model.hh#L85) attached to it:

```c++
// see if _entity is a visual
if (_ecm.EntityHasComponentType(_entity, components::Visual::TypeId())
{
  // do something with the visual entity here
}

// see if _entity is a model
if (_ecm.EntityHasComponentType(_entity, components::Model::TypeId())
{
  // do something with the model entity here
}
```